_posts/2021-03-09-bezier-fitting.md

```diff
@@ -0,0 +1,196 @@
+
+The main reason the problem is so hard is a basic but little-known fact about cubic beziers: C-shaped beziers come in triplets of very similar shape but quite different parameters. As a result, in general when fitting some arbitrary source curve, there are three local minima. Approaches based on iterative approximation (in addition to being slow) are likely to find just one, potentially missing a closer fit of one of the others.
+
+## How close are two curves?
+
+Before getting into the solution, we'll need to state the problem more carefully. The goal is a cubic bezier that minimizes some error metric with respect to the source curve. But how do we measure the distance between two curves?
+
+One meaningful metric is the [Fréchet distance]. It can be explained with a little story. An aircraft engineer goes on a walk with their dog Pierre, and the path followed by each can be described by a curve. What length leash is required so that the dog is always within that distance of the engineer? That is the Fréchet distance.
+
+Also note that the Fréchet distance is closely related to the Hausdorff distance, but the two can differ when the path loops back or crosses itself. For stroked curves, the Hausdorff distance is more relevant, but for filled paths such as font outlines, Fréchet preserves winding number in a way that Hausdorff doesn't guarantee.
+
+For solving optimization problems, the Fréchet distance is not ideal, as it effectively measures the distance of a single point, the maximum error. One can informally describe an [L2] error metric as similar, except that instead of describing the maximum length of the leash, it describes an *effort* of handling the dog which is proportional to the square of the distance of the dog from the engineer. Minimizing an L2 error means minimizing that total effort. This error metric takes into account the entire path, and thus is smoother in response to small changes of the parameters.
+
+I believe there is no one perfect error metric, and the choice depends on the context. Fortunately, for smooth curves such as we're likely to find in fonts, the differences between these error metrics are subtle, and we can confidently choose the one that's easiest to reason about mathematically. For this blog post, that's the L2 norm.
+
+Two things more to say: for smooth curves + low error, leash is generally perpendicular. Also, normalized arc length (both engineer and Pierre move at a constant speed) is a good approximation, though strictly larger.
+
+## Stating the problem
+
+Our dog Pierre has a funny quirk: he is capable of moving only along a cubic bezier path. That said, he is otherwise fairly obedient, and always starts and ends at the same position as the engineer, as well as starting out and ending up in the same direction.
```

Comment:
what is "position"? I presume it is not "point", but that is how I read it.

Reply:
"the same position relative to the engineer"?

Reply:
I think I've got it now but feel free to criticize the new version (which I'll commit soon).